### PR TITLE
Shadow Performance Improvements

### PR DESCRIPTION
**Pull request by unity-cchu**
### Checklist for PR maker
- [ ] Have you updated the changelog? Each package has a `CHANGELOG.md` file.

---
### Purpose of this PR
Refactor of shadow code. 
Now renders shadows to a single channel.
Prerendering of shadows can render to different channels in addition to different textures.
Added versioning
Added per light shadow culling
Made shader naming more consistent

### CODE REVIEW

File: com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowCaster2D.cs [hunk 1]
```diff
@@ -75,6 +85,17 @@ static int[] SetDefaultSortingLayers()
...
 
+        internal bool IsLit(Light2D light)
+        {
+            Vector3 deltaPos = light.transform.position - (m_ProjectedBoundingSphere.position + transform.position);
+            float distanceSq = Vector3.SqrMagnitude(deltaPos);
+
+            float lightRadiusSq = light.boundingSphere.radius * light.boundingSphere.radius;
+            float projectedRadiusSq = m_ProjectedBoundingSphere.radius * m_ProjectedBoundingSphere.radius;
+
+            return distanceSq <= (lightRadiusSq + projectedRadiusSq);
```
Comment (rolandk-unity) [inline com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowCaster2D.cs:96]: Is this correct? I guess `a^2 <= b^2 + c^2` is not the same as `a <= b+c`

File: com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs [hunk 2]
```diff
@@ -32,20 +44,80 @@ public static void InitializeBudget(uint maxTextureCount)
             }
         }
 
+        private static Material GetProjectedShadowMaterial(this Renderer2DData rendererData, int colorMask)
+        {
+            //if (rendererData.projectedShadowMaterial == null || rendererData.projectedShadowShader != rendererData.projectedShadowMaterial.shader)
```
Comment (rolandk-unity) [inline com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs:49]: Just double checking, do we want to create a new one every time the method is called? If so, maybe renaming to `CreateProjectedShadowMaterial` could make sense?

File: com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs [hunk 3]
```diff
@@ -32,20 +44,80 @@ public static void InitializeBudget(uint maxTextureCount)
...
+                rendererData.projectedShadowMaterial = material;
+            }
+
+            return rendererData.projectedShadowMaterial;
+        }
+
+        private static Material GetStencilOnlyShadowMaterial(this Renderer2DData rendererData, int colorMask)
+        {
+            //if (rendererData.stencilOnlyShadowMaterial == null || rendererData.projectedShadowShader != rendererData.stencilOnlyShadowMaterial.shader)
+            {
```
Comment (rolandk-unity) [inline com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs:63]: See `GetProjectedShadowMaterial` comment

File: com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs [hunk 4]
```diff
@@ -32,20 +44,80 @@ public static void InitializeBudget(uint maxTextureCount)
...
+                rendererData.stencilOnlyShadowMaterial = material;
+            }
+
+            return rendererData.stencilOnlyShadowMaterial;
+        }
+
+        private static Material GetSpriteSelfShadowMaterial(this Renderer2DData rendererData, int colorMask)
+        {
+            //if (rendererData.spriteSelfShadowMaterial == null || rendererData.spriteShadowShader != rendererData.spriteSelfShadowMaterial.shader)
+            {
```
Comment (rolandk-unity) [inline com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs:76]: See GetProjectedShadowMaterial comment

File: com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs [hunk 5]
```diff
@@ -32,20 +44,80 @@ public static void InitializeBudget(uint maxTextureCount)
...
+                rendererData.spriteSelfShadowMaterial = material;
+            }
+
+            return rendererData.spriteSelfShadowMaterial;
+        }
+
+        private static Material GetSpriteUnshadowMaterial(this Renderer2DData rendererData, int colorMask)
+        {
+            //if (rendererData.spriteUnshadowMaterial == null || rendererData.spriteUnshadowShader != rendererData.spriteUnshadowMaterial.shader)
+            {
```
Comment (rolandk-unity) [inline com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs:88]: See GetProjectedShadowMaterial comment

**Review by rolandk-unity — APPROVED** [review]
Looks good overall

File: com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs (hunk 2 above)
Comment (unity-cchu) [reply com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs]: Fixed

File: com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs (hunk 3 above)
Comment (unity-cchu) [reply com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs:83]: Fixed

File: com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs (hunk 4 above)
Comment (unity-cchu) [reply com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs:94]: Fixed

File: com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs (hunk 5 above)
Comment (unity-cchu) [reply com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowRendering.cs:105]: Fixed

File: com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowCaster2D.cs (hunk 1 above)
Comment (unity-cchu) [reply com.unity.render-pipelines.universal/Runtime/2D/Shadows/ShadowCaster2D.cs:96]: Good catch

**Review by phi-lira — APPROVED** [review]

**Comment by tatoforever** [comment]
Folks, why static meshes with baked shadows must also render to shadowmaps in order to receive shadows? Isn't this a bit pointless? What about "let any mesh receive shadows", no matter if they cast shadows or not? Could improve performance as you don't need to render all geometry reciving shadows. I have a custom renderer that sit on top of built-in that ouperforms URP by a large marge (2x - 3x easily) when there's many lights casting shadows cause none of my static geometry renders to depthmaps.

**Pull request merged**